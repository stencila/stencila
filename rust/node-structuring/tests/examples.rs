use std::{
    fs::{read_dir, read_to_string},
    path::PathBuf,
    str::FromStr,
};

use eyre::{OptionExt, Result, bail};
use insta::assert_json_snapshot;
use itertools::Itertools;

use stencila_codec::{Codec, StructuringOperation, StructuringOptions, stencila_format::Format};
use stencila_codec_json::JsonCodec;
use stencila_codec_markdown::MarkdownCodec;
use stencila_node_structuring::structuring;
use stencila_schema::{Node, Primitive};

/// Structure example Markdown or JSON files using the structuring operation
/// specified in the first three parts of their name, or multiple operations for
/// files with numeric prefixes.
///
/// Run using:
///
/// cargo insta test -p stencila-node-structuring
///
/// and then review snapshots with:
///
/// cargo insta review
#[tokio::test]
async fn examples() -> Result<()> {
    let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/examples");

    for entry in read_dir(&dir)?.flatten() {
        let path = entry.path();
        let format = Format::from_path(&path);

        let content = read_to_string(&path)?;

        let (mut node, ..) = match format {
            Format::Markdown => MarkdownCodec.from_str(&content, None).await?,
            Format::Json => JsonCodec.from_str(&content, None).await?,
            _ => bail!("Unsupported formatted"),
        };

        let id = path
            .file_stem()
            .map(|name| name.to_string_lossy().to_string())
            .ok_or_eyre("No file stem")?;

        let Node::Article(article) = &node else {
            bail!("Expected an article")
        };

        let operations = if let Some(Primitive::Array(ops)) = article
            .options
            .extra
            .as_ref()
            .and_then(|extra| extra.get("structuring").cloned())
        {
            ops.iter()
                .filter_map(|op| match op {
                    Primitive::String(op) => StructuringOperation::from_str(op).ok(),
                    _ => None,
                })
                .collect_vec()
        } else {
            // Fall back to operation from filename for single operation tests
            let op = id.split("-").take(3).collect_vec().join("-");
            if let Ok(operation) = StructuringOperation::from_str(&op) {
                vec![operation]
            } else {
                bail!("Unable to determine structuring operations for {}", path.display());
            }
        };

        structuring(&mut node, StructuringOptions::new(operations, []));

        assert_json_snapshot!(id, node);
    }

    Ok(())
}
