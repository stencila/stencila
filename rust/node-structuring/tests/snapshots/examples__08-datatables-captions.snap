---
source: rust/node-structuring/tests/examples.rs
expression: node
---
{
  "type": "Article",
  "frontmatter": "structuring:\n- tables-to-datatables\n- tables-with-captions",
  "content": [
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "This tests that simple tables with captions are properly converted to datatables while preserving caption association. This interaction is critical for structured data with descriptions - the caption must be correctly associated with the datatable, and only uniform simple tables should be converted."
          }
        }
      ]
    },
    {
      "type": "Datatable",
      "id": "tab-1",
      "label": "1",
      "labelAutomatically": false,
      "caption": [
        {
          "type": "Paragraph",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "Performance metrics showing processing times and accuracy rates."
              }
            }
          ]
        }
      ],
      "columns": [
        {
          "type": "DatatableColumn",
          "name": "Method",
          "values": [
            "A",
            "B",
            "C"
          ],
          "validator": {
            "type": "ArrayValidator",
            "itemsValidator": {
              "type": "StringValidator"
            }
          }
        },
        {
          "type": "DatatableColumn",
          "name": "Time (ms)",
          "values": [
            120,
            85,
            203
          ],
          "validator": {
            "type": "ArrayValidator",
            "itemsValidator": {
              "type": "IntegerValidator"
            }
          }
        },
        {
          "type": "DatatableColumn",
          "name": "Accuracy",
          "values": [
            95.2,
            97.1,
            89.7
          ],
          "validator": {
            "type": "ArrayValidator",
            "itemsValidator": {
              "type": "NumberValidator"
            }
          }
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "The simple table above should be converted to a datatable because it has consistent rows, simple text-only cells, and no spans. The caption should be preserved and associated with the resulting datatable."
          }
        }
      ]
    }
  ],
  "structuring": [
    "tables-to-datatables",
    "tables-with-captions"
  ]
}
